styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java

```diff
@@ -183,6 +190,75 @@
     this.forkJoinPool = new ForkJoinPool(REQUEST_CONCURRENCY);
   }
 
+  /**
+   * Bootstrap active worflow instance index. This cannot run concurrently with the scheduler and must only
+   * be called once before the scheduler starts.
+   */
+  void indexActiveWorkflowInstances() {
+    // XXX: this listing is not strongly consistent
+    final List<Entity> indexEntries = StreamUtil.stream(
+        datastore.run(Query.newKeyQueryBuilder().setKind(KIND_ACTIVE_WORKFLOW_INSTANCE).build()))
+        .map(Key::getName)
+        .map(wfiKey -> activeWorkflowInstanceIndexShardEntryKey(datastore.newKeyFactory(), wfiKey))
+        .map(indexEntryKey -> Entity.newBuilder(indexEntryKey).build())
+        .collect(toList());
+
+    for (List<Entity> batch : Lists.partition(indexEntries, MAX_NUMBER_OF_ENTITIES_IN_ONE_BATCH)) {
+      datastore.put(batch.toArray(new Entity[0]));
+    }
+  }
+
+  /**
+   * This indexing can run concurrently with the scheduler operations as it is using transactions to ensure it
+   * consistently updates the index, but that also makes it too slow to do the initial bootstrap.
+   *
+   */
+  void concurrentlyIndexActiveWorkflowInstances() {
+    // Weak listing, but we'll eventually read _everything_, in due time...
+    final List<Entity> expectedIndexEntries = StreamUtil.stream(
+        datastore.run(Query.newKeyQueryBuilder().setKind(KIND_ACTIVE_WORKFLOW_INSTANCE).build()))
+        .map(Key::getName)
+        .map(wfiKey -> activeWorkflowInstanceIndexShardEntryKey(datastore.newKeyFactory(), wfiKey))
+        .map(indexEntryKey -> Entity.newBuilder(indexEntryKey).build())
+        .collect(toList());
+
+    // Fetch index entries in batches
+    for (List<Entity> expectedEntryBatch : Lists.partition(expectedIndexEntries, MAX_NUMBER_OF_ENTITIES_IN_ONE_BATCH)) {
+
+      final List<Entity> actualEntryBatch = datastore.fetch(expectedEntryBatch.stream()
+          .map(Entity::getKey).toArray(Key[]::new));
+      assert actualEntryBatch.size() == expectedEntryBatch.size();
```

Comment:
Is this intended? Isn't assertion disabled by default at runtime?

Reply:
Can change this to `if/throw` check instead.

Reply:
`assert` is probably good enough. I'm expecting this thing to fail.

Reply:
If you're expecting it to fail then we should probably make it a real `if` check =)